File: IAIso-v5.0/core/iaiso-csharp/tests/Iaiso.Tests/MiddlewareTests.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Audit;
using Iaiso.Core;
using Iaiso.Middleware;
using Iaiso.Middleware.Anthropic;

namespace Iaiso.Tests;

public sealed class MiddlewareTests
{
    private sealed class FakeAnthropicClient : IAnthropicClient
    {
        private readonly AnthropicResponse _resp;
        public FakeAnthropicClient(AnthropicResponse resp) { _resp = resp; }
        public AnthropicResponse MessagesCreate(JsonObject p) => _resp;
    }

    public void TestAnthropicAccountsTokensAndToolCalls()
    {
        var sink = new MemorySink();
        using var exec = BoundedExecution.Start(new BoundedExecutionOptions { AuditSink = sink });

        var raw = new FakeAnthropicClient(new AnthropicResponse(
            "claude-opus-4-7", 100, 250,
            new List<ContentBlock>
            {
                new("text"),
                new("tool_use"),
                new("tool_use"),
            }));
        var client = new BoundedAnthropicClient(raw, exec);
        client.MessagesCreate(new JsonObject());

        bool foundStep = false;
        foreach (var e in sink.Events())
        {
            if (e.Kind == "engine.step")
            {
                Assert.Equal(350L, (long)e.Data["tokens"]!);
                Assert.Equal(2L, (long)e.Data["tool_calls"]!);
                foundStep = true;
            }
        }
        Assert.True(foundStep, "expected engine.step event");
    }

    public void TestAnthropicRaisesOnEscalationWhenOptedIn()
    {
        var cfg = PressureConfig.CreateBuilder()
            .WithEscalationThreshold(0.4)
            .WithReleaseThreshold(0.95)
            .WithDepthCoefficient(0.5)
            .WithDissipationPerStep(0.0)
            .Build();
        using var exec = BoundedExecution.Start(new BoundedExecutionOptions { Config = cfg });
        // Force escalation
        exec.RecordStep(new StepInput { Depth = 1 });

        var raw = new FakeAnthropicClient(new AnthropicResponse("model", 0, 0, null));
        var client = new BoundedAnthropicClient(raw, exec, new AnthropicOptions { RaiseOnEscalation = true });
        Assert.Throws<MiddlewareException.EscalationRaised>(() =>
            client.MessagesCreate(new JsonObject()));
    }
}

File: IAIso-v5.0/core/iaiso-swift/Tests/IAIsoMiddlewareTests/MiddlewareTests.swift
import XCTest
@testable import IAIsoAudit
@testable import IAIsoCore
@testable import IAIsoMiddleware

final class MiddlewareTests: XCTestCase {

    final class FakeAnthropic: Anthropic.Client {
        func messagesCreate(_ params: [String: Any]) throws -> Anthropic.Response {
            return Anthropic.Response(
                model: "claude-opus-4-7",
                inputTokens: 100, outputTokens: 250,
                content: [
                    Anthropic.ContentBlock(type: "text"),
                    Anthropic.ContentBlock(type: "tool_use"),
                    Anthropic.ContentBlock(type: "tool_use"),
                ])
        }
    }

    func testAccountsTokensAndToolCalls() throws {
        let sink = MemorySink()
        let exec = try BoundedExecution.start(.init(auditSink: sink))
        let client = Anthropic.BoundedClient(
            raw: FakeAnthropic(), execution: exec, options: .defaults)
        _ = try client.messagesCreate([:])

        var foundStep = false
        for e in sink.events where e.kind == "engine.step" {
            if let t = e.data["tokens"]?.intValue, t == 350,
               let tc = e.data["tool_calls"]?.intValue, tc == 2 {
                foundStep = true
            }
        }
        XCTAssertTrue(foundStep)
        exec.close()
    }

    func testRaisesOnEscalationWhenOptedIn() throws {
        let cfg = PressureConfig(
            depthCoefficient: 0.5,
            dissipationPerStep: 0.0,
            escalationThreshold: 0.4,
            releaseThreshold: 0.95)
        let exec = try BoundedExecution.start(.init(config: cfg))
        _ = exec.recordStep(StepInput(depth: 1)) // force escalation

        let client = Anthropic.BoundedClient(
            raw: FakeAnthropic(), execution: exec,
            options: Anthropic.Options(raiseOnEscalation: true))
        XCTAssertThrowsError(try client.messagesCreate([:])) { err in
            if case MiddlewareError.escalationRaised = err {
                // expected
            } else {
                XCTFail("expected escalationRaised, got \(err)")
            }
        }
    }
}
